index.rst

```diff
@@ -1,29 +1,192 @@
+This tech note proposes a design for implementing user impersonation in Gafaelfawr_, the authentication system for the Science Platform.
+It focuses on user impersonation to services used via the browser, such as the Notebook Aspect and the Portal Aspect.
+Impersonation of users to API services using bearer tokens is already possible by creating a new token for that user.
+This will not be changed by this proposal.
+
+.. _Gafaelfawr: https://gafaelfawr.lsst.io/
 
 .. note::
 
-   **This technote is a work-in-progress.**
+   This is part of a tech note series on identity management for the Rubin Science Platform.
+   The primary documents are DMTN-234_, which describes the high-level design; DMTN-224_, which describes the implementation; and SQR-069_, which provides a history and analysis of the decisions underlying the design and implementation.
+   See the `references section of DMTN-224 <https://dmtn-224.lsst.io/#references>`__ for a complete list of related documents.
 
-Abstract
-========
+.. _DMTN-234: https://dmtn-234.lsst.io/
+.. _DMTN-224: https://dmtn-224.lsst.io/
+.. _SQR-069: https://sqr-069.lsst.io/
+
+Terminology
+===========
+
+For the purposes of this tech note, an *administrator* is a user who is flagged as an administrator in Gafaelfawr_.
+Those users are automatically granted the ``admin:token`` scope for their session token when they log in.
+
+An *impersonated user* is the apparent user from the perspective of the service.
+This is the user that the administrator is temporarily pretending to be.
+
+The *impersonator* is the administrator doing the impersonation.
+
+The *impersonation token* is the additional token that the impersonator obtains and uses during impersonation.
+This token will be used as the basis for authentication and access control while impersonation is enabled.
+
+Requirements
+============
+
+It is already possible, using a token with ``admin:scope``, to create a user token for an arbitrary user.
+This allows impersonation of arbitrary users to API services authenticated with bearer tokens.
+However, it is not possible for an administrator to put such a token in a browser cookie, and it is therefore difficult to use such tokens for access to browser-based services such as the Notebook Aspect or the Portal Aspect.
+
+Via a UI in a web browser, administrators should be able to start impersonating an arbitrary user.
+While that impersonation is active, any service they visit in that browser session should believe that the administrator is the impersonated user in every respect except (optionally) logging and a UI display.
+For example, new notebook lab pods should be launched with the impersonated user's UID, GID, and groups; TAP queries should be saved in the impersonated user's history; all Portal actions should be done as the impersonated user; and so forth.
+
+This impersonation should honor the following constraints:
+
+#. ``admin:token`` scope is required to impersonate another user.
```

Comment:
Is there any worry about an inside attack scenario with a support person using `admin:token` to gain privileges by impersonating someone with `admin:provision` or `exec:admin`?

Like, should we have a mechanism to ensure that a person can't gain additional scopes through impersonation? Or maybe we merely have an extra flag in the Slack alerts that the impersonation is elevating privileges?

Reply:
This is okay in the current security model. `admin:token` is full root; it grants you access to do absolutely anything protected by Gafaelfawr, since you can use it to create arbitrary tokens for any user with any scope. The changes in this tech note therefore don't increase its powers.

We could potentially split things up so that there is a model of impersonation where you're only allowed to impersonate non-admin users and it is authorized by a different scope than the "create literally any token" scope, but I think that may be unnecessary complexity. My expectation is that any given deployment of the RSP is going to have on the order of five admins who are running the whole thing anyway and will have absolute powers, and those are the only people who need to do user impersonation.

This would change if we decide that more front-line support people need user impersonation, which may be a good discussion to have with Frossie about what her expectations are.